### ISSUE TITLE
Restore event log replay functionality.

### ISSUE THREAD

**Issue by sigilante** [urbit/vere#157]
> [Pull] the event log replay functionality out of v1.14 of Vere with the intention of it landing in v1.19.

> 1.14 was reverted due to some kind of corruption error but the error was from a different part of the release so the idea here is to extract the replay improvements from 1.14 and make a PR that just contains those

**Comment by sigilante** [comment]
> None of the meld/memory paging/printing stuff, just the replay.
>
> Not this, for instance: https://github.com/urbit/urbit/pull/6159/files#diff-9b3877485f99b0b6591908d549b6ac16cf7e327e33d941018ca339e590622a47R648
>
>Hmm, it seems to rely on automeld. Moving convo to issue.

`_cw_play` in `main.c` uses `u3o_auto_meld` and thus implicates the whole meld set of changes.

**Comment by sigilante** [comment]
@belisarius222 what was actually causing the memory corruption issue?  It wasn't the meld stuff was it?

**Comment by joemfb** [comment]
Automeld was added (ported) for that feature, specifically. On that branch / as of that pr, it wasn't used anywhere else.

**Comment by sigilante** [comment]
What I'm worried about is that I'm going to inadvertently regress something. There are a lot of data structures with single elements added because of all the automeld and CLI options.

Git diffs don't work because the repo rearchitecture changed so many minor things as to be unwieldy.

Thoughts?

Example: the code I'm adding back in refers to gar_w in _u3e_pool.

https://github.com/urbit/urbit/blob/a7880717e3c77c21e185ffe52a4982b0b3bd52d1/pkg/urbit/include/noun/events.h#L45

So I have to add this back in, or I have to backtrace the entire logic for why it's not there now.

Include all of the guard page stuff as well then?

**Comment by sigilante** [comment]
Here's the direct comparison:  https://github.com/urbit/vere/compare/urbit-v1.14-rc3...vere-v1.17

I've been working off of https://github.com/urbit/urbit/pull/6159/files to make changes thus far.

**Comment by matthew-levan** [comment]
@sigilante what's the difference between the functionality you're referring to here and the functionality provided by the existing `--replay-from NUMBER`, `--replay-to NUMBER`, and `--replay` options?

**Comment by sigilante** [comment]
@matthew-levan This restores the 1.14 replay stuff:

> - adds the `play` command, for improved replay of large event logs,
with automatic retry of bail:meme after pack (or meld with --auto-meld),
snapshot updates after errors or interruption (via ctrl-c), and
optional full replay (without editing files in the pier).

**Comment by mcevoypeter** [comment]
The `git diff` troubles should no longer exist; when we migrated repos, I ran `git filter-repo` on https://github.com/urbit/urbit to filter out non-runtime commits, which obviously rewrote a ton of commits. I forgot to push the tags, but I just pushed `urbit-v1.14-rc3` now, which yields a much saner, 84-commit diff for you to work off of. I'll push the remaining `urbit-*` tags soon.

**Comment by sigilante** [comment]
_Much_ more manageable than the 600+ files, 90,000+ additions I was trying to filter against!

**Comment by sigilante** [comment]
To confirm then: I do need to add back in the guard page stuff, since that is part of the changes the `play` command relies on.  @mcevoypeter

**Comment by sigilante** [comment]
@belisarius222 says the guard page stuff isn't in scope, so I'll prepare a quick summary of where it's currently referred to in the `play` work and post for comment.

**Comment by sigilante** [comment]
Basically:

- `u3_mars_play` relies on `u3m_save`
- `u3m_save` refers to `u3e_fault`
- `u3e_fault` has a new prototype and refers as well to `u3P.gar_w`, the guard page
- `gar_w` is added in `events.h`

The simplest solution is to make `u3_mars_play` not use `u3m_save` if possible.

**Comment by joemfb** [comment]
Just call `u3e_save()` instead of `u3m_save()`.

**Comment by sigilante** [comment]
The current status is that the branch code correctly produces working fake ships etc.

Replay doesn't error but appears to hang.  No event replay logging is visible.  There's a `_term_it_refresh_line` function that I can use but isn't implemented in current `develop`.

The ship cannot be terminated in replay mode.  This may be due to terminal changes in Ctrl+C/Ctrl+Z handling.

```
./bazel-bin/pkg/vere/urbit play zode
loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 652 jets
--------------- bootstrap starting ----------------

boot: 1-4

--------------- bootstrap complete ----------------

---------------- playback starting ----------------

play: events 5-123
```

**Comment by sigilante** [comment]
OK, that's a complete review against the 1.14 codebase, and removal of some extra demand paging bits that I leaked in.

Still hangs on the actual replay tho.

**Comment by sigilante** [comment]
Huh.  Deleting the pier results in:
1. Regular behavior when the ship is booted (replay events)
2. Hanging behavior when the ship is just `urbit play`ed

**Comment by sigilante** [comment]
It's playing some events (bootstrap) but not others???

**Comment by joemfb** [comment]
What's does a debugger say it's doing when it's hanging? Can you get a backtrace?

**Comment by sigilante** [comment]
It's cycling between `allocate.c` (`u3a_celloc`) and `hashtable.c` (`_ch_node_git`).  Ted is going to look over it with me this afternoon.

**Comment by mcevoypeter** [comment]
Are you on macOS? If so, there's an issue with `-O3 -g` on macOS that leads to a hang that sounds similar to yours. [b16ada6baaccdce2a3dda78caf70b55d8dbbc28e](https://github.com/urbit/vere/tree/b16ada6baaccdce2a3dda78caf70b55d8dbbc28e) removes `-g` when compiling on macOS.

**Comment by sigilante** [comment]
No, it's not, and it looks like it is in fact playing back events but _incredibly_ slowly.  (I left it for hours and it has the arvo kernel processing happening now.)

**Comment by joemfb** [comment]
Nock being suddenly, shockingly slow means that jets aren't firing. You're properly setting `u3o_hashless`, which is the first thing I would've checked. Much has changed, so I'm not sure where else to look off the top of my head.

**Comment by zalberico** [comment]
@sigilante if you're still stuck can you pair with @joemfb on this? Nearly all of the 1.14 commits were from Joe so he should know what's important to cherry pick over to get the replay bits of it up and running.

**Comment by sigilante** [comment]
I teach ASL tonight then I'll take a look at it first priority Thursday morning.  Ted and I got a lot of review through it and I'll follow up ASAP.

**Comment by sigilante** [comment]
Wait.  This is my vanilla `vere` build.

```
urbit 1.18-9500da18fd
boot: home is nec
loom: mapped 2048MB
lite: arvo formula 2a2274c9
lite: core 4bb376f0
lite: final state 4bb376f0
boot: loading pill /home/neal/urbit/vere/solid.pill
boot: parsing %solid pill
loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 652 jets
```

Note that last line:  there should be 351 jets.  So something's wrong with my local `bazel` build.  Switching to chat to follow up on whether this is general (and needs another issue) or a local misconfiguration.

**Comment by sigilante** [comment]
Obviously things "work"—ships are functional etc.—but since I'm trying to debug jets not firing this may be relevant.

**Comment by sigilante** [comment]
OK, I've narrowed down the slow patch to `u3m_soft_top`, line

```
  if ( 0 == (why = (u3_noun)_setjmp(u3R->esc.buf)) ) {
```

`_setjmp` slowing things up in the replay?

**Comment by sigilante** [comment]
Status as of now:

- default replay (not with `play` command) appears to be normal
- the `SIGINT` handler for `play` may not be registered somewhere it needs to be (Ctrl+Z doesn't actually kill it).
- the jet situation is ambiguous, but I do know that some jets (like `+add`) are firing

The blocker is analyzing why `_setjmp` is so slow, which is beyond my current C.

**Comment by sigilante** [comment]
Bigger picture:  why does `urbit play` look so different from `_pier_play()`, which is what triggers when the checkpoint is gone on a regular ship start?

**Comment by sigilante** [comment]
Following up on the scenarios:

1.  No checkpoint, slow replay.
2.  Any checkpoint, fast (regular) replay.
3.  Any checkpoint, forced (`play -f`) full replay, slow replay.

The place I checked next is in `_cw_play`, but `ful_o` isn't not set in one and set in the other, so I don't think it's this conditional changing anything.

```
  if ( c3y == ful_o ) {
    u3l_log("mars: preparing for full replay\r\n");
    u3m_init((size_t)1 << u3_Host.ops_u.lom_y);
    u3e_live(c3n, u3_Host.dir_c);
    u3e_yolo();
    u3m_pave(c3y);
    u3j_boot(c3y);
    u3A->eve_d = 0;
  }
  else {
    u3m_boot(u3_Host.dir_c, (size_t)1 << u3_Host.ops_u.lom_y);
  }
```

So what's making the difference in speed presumably arises from what's being done differently elsewhere.

The next place to focus in on is `u3_mars_play`.

**Comment by Fang-** [comment]
re terminal output shenanigans: as of v1.17 `u3l_log` does not require trailing `\r\n`. Adding those in anyway will result in extraneous newlines/whitespace. Shoot me a dm if you're still seeing anomalous output outside of that. You shouldn't need to call any of the `_term_*` functions explicitly.

**Comment by sigilante** [comment]
@Fang- since I may as well clean up some code while I have this pulled up, is there a reason to not change failure messages from `fprintf` to `u3l_log`?

For instance,

```
      fprintf(stderr, "mars: disk read meta fail\r\n");
```

**Comment by Fang-** [comment]
Pretty sure everything that uses `fprintf` happens before the IO drivers have been initialized. `u3l_log` depends on the `term.c` driver.

`u3l_log` may fall back to doing `fprintf` _anyway_ if the driver hasn't been initialized yet, but I'd just leave existing calls for what they are. There might be another reason why they're explicit `fprintf`s that I'm not aware of. (And the changes would be outside the scope of this PR anyway.)

**Comment by sigilante** [comment]
~master-morzod and I went through this. He suspects that either the compiler jets are not firing for some reason or the hash table (memoization) "incomprehensibly slower." It's not clear why this would be the case given what's actually in this PR—it should either work or break in an obvious way. "Something strange is happening here."

**Comment by zalberico** [comment]
@sigilante can you write up exact repro steps here for @mcevoypeter ? He'll take a look on Monday.

**Comment by sigilante** [comment]
1. Clone the current `i/157` branch.
2. Build it normally.
3. Boot a fakezod.
4. Scenario A:  rm -rf zod/.urb/chk/* && urbit zod—this should be a normal playback.
5. Scenario B:  rm -rf zod/.urb/chk/* && urbit play zod—this is very slow, and it seems to be in the compiler code.
6. Scenario C:  mv zod/.urb/chk/* zod/.urb/bhk, then boot the ship and do some things, then mv zod/.urb/bhk/* zod/.urb/chk && urbit play zod—this is a normal playback.
7. Scenario D:  As C except urbit play -f zod—this is a very slow playback.

Scenarios B and D are the odd ones.

**Comment by mcevoypeter** [comment]
Thanks, @sigilante. After playing around with this this morning, I'm not seeing anything that makes me think it's a build system issue, especially because normal replay works as expected. I'm going to comb through the diff between this branch and `develop` to see if anything jumps out at me and will update here later today.

**Comment by sigilante** [comment]
I let the fakezod run the slow version all weekend, and it got through event 10.  I think it is the compiler jets failing to fire (for whatever reason).

**Comment by zalberico** [comment]
Update from the sync, @barter-simsum is testing replay from live -> release to make sure regular replay still works post doccords.
- Boot a comet
- Do some stuff, delete the chk, run a replay (test using what's currently on the live network).
- Confirm that works
- Install `soon` channel binary via next (put soon in pace file in `.bin` and `.run next`)
- `|ota ~marnec-dozzod-marzod` - this will install the pre-release 415 arvo
- Delete the comet's `chk`
- Run a replay
- Confirm that works

The goal is to confirm regular replay is fine post doccords (and is working on the current live network too).

**Comment by barter-simsum** [comment]
Replay works. Vere binary pulled from `edge` rather than `soon` and `soon` didn't seem to support a high enough kelvin version.

**Comment by zalberico** [comment]
Sweet - thanks for testing, we're good for tomorrow's release then. The issue seems isolated to this PR's code then which we can debug. If it's not figured out today it won't make this release for Feb 7th. If it merges sometime prior to Feb 7th then it'll get into the branched release with a target ship date of Feb 14th.

**Comment by zalberico** [comment]
I have an idea that'd be interesting:
- Cherry pick the 1.14 replay code into 1.15 and build it with the old build system
- Cherry pick the 1.14 replay code into 1.16 and build it with the new build system
- 1.15 and 1.16 are code equivalent, only the build system is different
- Compare the results

Basically is the weird jet behavior something about this PR and 1.15 or something about the build.

**Comment by sigilante** [comment]
Yes, that's an interesting idea.  I'll try to play with that during the release call tomorrow.

**Comment by mcevoypeter** [comment]
I spent a good chunk of today looking into why `urbit play <pier>` runs so slowly on `i/157/event-log-replay-restoration`. I couldn't find any build-system related explanation, nor could I find any errors related to migration from the old repo to the new one. I do think @zalberico's idea is a good one to gain further insight into this bug.

As a result of combing through the diff for this feature, I realized that we can implement an improved event log replay with much less additional complexity than the version that shipped with v1.14 at the expense of retrying events that fail because they run out of memory i.e. bail meme.

We can easily use the existing replay implementation--the one a ship goes through at launch--to implement the `play` subcommand and simply exit successfully right after replay finishes. After each replay batch succeeds, we'll take a snapshot. We can also allow the user to specify the batch size for replay from the command line, which will enable users to replay as few as a single event at a time. The combination of snapshotting after each successful replay batch completion and configurable replay batch size is still a major improvement.

I have a working implementation of this, albeit without the command line configurable batch size, but that won't take much effort to implement.

**Comment by joemfb** [comment]
@sigilante @Fang- `fprintf()` is preferable for error messages, because the `u3l_log()` implementation can be much more complex (async in the presence of IPC) and therefore less reliable. This has been a support obstacle in the past.

@mcevoypeter the improved error handling and event retry was the main reason for adding a separate replay command. It was only separate because mars/urth isn't shippable -- it's preferable for replay to be automatic in almost all cases. Snapshot updates on error/exit would definitely be an improvement, but doing so after every batch will result in a very high `write()` volume (unless batch size is increased massively, which will likely have performance problems). Async replay is just much more difficult to control.

I'll take another turn with this PR in a debugger as soon as I can.

**Comment by mcevoypeter** [comment]
I understand the value that the improved error handling and event retry brings, but the simple, piggybacking approach I detailed above still does add value. In the interest of delivering an improved replay experience to users in the short term, I'd argue we should ship the piggybacking approach while we figure out what's going on with this the more full-featured version. I implemented the command line arguments, and, all told, the diff is ~80 lines. If we squash and merge the resulting PR, it'll also be trivial to revert should we want to do so in the future.

**Comment by sigilante** [comment]
I have been comparing jet invocations in `u3j_kink` for the two sequences.

The regular replay (boot and replay, the fast existing one) and the slow replay (`urbit play`, this PR) appear to be basically identical, with two more Nock invocations in the slow replay.

The full output is way too long for this PR discussion, but the relevant portion (at event 5 onwards), looks like this:

"regular":

```
$u3j_kink nock

$u3j_kink nock

$u3j_kink nock

$u3j_kink nock
$u3j_kink nock
$u3j_kink nock


*u3j_kink jet
*u3j_kink jet
$u3j_kink nock
...
```

slow:

```
$u3j_kink nock

$u3j_kink nock

$u3j_kink nock

$u3j_kink nock
$u3j_kink nock
$u3j_kink nock
$u3j_kink nock
$u3j_kink nock
*u3j_kink jet
*u3j_kink jet
$u3j_kink nock
...
```

Everything before the event replay is identical, hundreds of lines' worth.  This tends to undermine the hypothesis that compiler jets aren't firing.

**Comment by philipcmonk** [comment]
I think u3j_kink looks up jets, right?  It being fired doesn't mean the jets were actually found, you'd need to see which branch it took

**Comment by sigilante** [comment]
Right, I'm printing nock/jet in the conditional after they're found.

Anyway, it looks like my replay is locally slow anyway in both cases right now, so the mystery deepens for me.

**Comment by zalberico** [comment]
https://github.com/urbit/vere/issues/157#issuecomment-1409485099

@sigilante were you able to try that?

**Comment by zalberico** [comment]
> I have an idea that'd be interesting:
> 
> * Cherry pick the 1.14 replay code into 1.15 and build it with the old build system
> * Cherry pick the 1.14 replay code into 1.16 and build it with the new build system
> * 1.15 and 1.16 are code equivalent, only the build system is different
> * Compare the results
> 
> Basically is the weird jet behavior something about this PR and 1.15 (code equivalent to 1.13 and 1.16) or something about the build (since we never tested this code outside of 1.14, that's the other variable - this test would reduce the variable down to only the different build system which is the only thing different between 1.15 and 1.16 so we at least know for sure if it's that or not).

@sigilante were you able to try that?

**Comment by sigilante** [comment]
I didn't do it against 1.15 and 1.16 yet—just against an updated `develop`.  I'll give it a shot.

**Comment by zalberico** [comment]
Reopening - it got automatically closed accidentally by github due to a referenced issue (the simple replay command fix)

**Comment by sigilante** [comment]
1. Trying to build the old versions in the monorepo is now a phenomenal pain.
2. I have been unable to get v1.15 with these changes to compile due to header/linker stuff, will play with it more.
3. I can get v1.16 with these changes to compile and parse a pill, but the Dojo will not load.

**Comment by sigilante** [comment]
Have a gander at #210 for a fix.

It clobbered some changes that I need to restore but it plays at regular speed.

**Comment by joemfb** [comment]
This is really bizarre.

There's still a big diff between those two PRs, due to many other changes throughout the repo. It tried merging `develop` back into each branch. Other than whitespace, the only differences I see are the order of header includes in `main.c`, and the set of headers included in `mars.c`. Header conventions were changed in a big way while replacing the build and exiting the monorepo. But I'm quite surprised to see this causing behavior changes or bugs like this instead of just failing to build.

**Comment by sigilante** [comment]
`#includes` and which directory files are located are the big differences.

**Comment by sigilante** [comment]
The #210 PR's diff include some code dropped from `serf.c` (`_serf_space` and `_serf_prof``, a `usage_c` error message in `main.c`, and in `disk.`c some code in `_disk_read_one_cb` and `_disk_serialize_v1`.  What should I do with that, as I'm not sure what it's for offhand?

**Comment by joemfb** [comment]
@sigilante I left a review on your pr requesting some changes. But all the things you mentioned here are good and should be kept.

**Comment by mcevoypeter** [comment]
@sigilante can you be more specific? Which blocks of `#include`s changed and in which files? And what do you mean exactly by "where directory files are located"? It'd be great if we could understand what single change causes this work to break.

**Comment by joemfb** [comment]
@mcevoypeter here's the diff of the merges i mentioned above: https://pastebin.com/YPnG7kF8

**Comment by sigilante** [comment]
- If I build a fakezod from the 1.15 solid.pill, then `urbit play` replay is fast.
- If I build a fakezod from the 1.19 solid.pill, then `urbit play` replay is slow.

Based on this, I'm going to say that `urbit play` and the associated changes are not the problem, they are a symptom of something else (maybe urth/mars split?) going wrong.

**Comment by matthew-levan** [comment]
To add on what @sigilante said, it is also slow on fakezods freshly booted with `urbit-1.17.pill` and `urbit-1.18.pill` pills (both from bootstrap.urbit.org).

**Comment by zalberico** [comment]
I'm confused - I thought we already confirmed that the performance is fast with the changes in the diff Joe linked here?: https://pastebin.com/YPnG7kF8 - and this diff was the cause (for some as yet unknown reason). If that's the case we'd expect 1.15 to be fast and 1.19 to be slow (if it doesn't have the change in the diff).

And we have a working change in #210 

Are you saying that's not the case and #210 is also slow?

**Comment by sigilante** [comment]
I'm saying it is a function of the pill used to boot Arvo.

**Comment by matthew-levan** [comment]
@zalberico, @sigilante was testing this code on a fakezod he had booted with a `1.15` pill this entire time, so we were under the impression that we had something working. However, I tried to test it on my machine last week and confirmed that it's slow on ships booted with any of the future pills (`1.16` *through* `1.19`).

**Comment by zalberico** [comment]
I understand, thanks and good catch debugging!

**Comment by sigilante** [comment]
Matt tested 1.16

**Comment by joemfb** [comment]
The pills alone can't explain the difference, if we're able to boot the ships but then not replay them with the new implementation). The only way I can explain that is in differences between jet state/dashboards (between `serf.c` and `mars.c`) -- I observed that above and can reproduce if needed. I still do not understand how the differing header includes can still compiles and runs, but produce different behavior. I believe that something about the way we are now declaring/initializing global variables is to blame.

**Comment by mcevoypeter** [comment]
The change in how we declare and define global variables is now standard: we now declare global variables as `extern` in the appropriate header file and define them in the appropriate source file corresponding. Before, we were defining global variables in header files, which causes multiple definition errors at link time (when using standard header inclusion practices).

I agree it's weird that the differing includes produce different behavior. However, it's also positive pressure since the behavior difference likely means we were getting away with something we shouldn't have been able to before the switch to the new build system (and accompany header inclusion pattern changes).

All that said, I noticed `u3j_Dash` is not defined in `pkg/noun/jets.c`, which it should be. However, after defining it and replaying with `play`, I'm not seeing any difference in behavior: event 7 still takes a prohibitively long time. It's also worth pointing out that `u3_Host` has no clear owning file, so its definition may be worth looking into.

**Comment by matthew-levan** [comment]
@joemfb pointed my attention back at L25 of `pkg/noun/BUILD.bazel`, which uses the `-force_load` option to resolve a linker error that otherwise occurs during build.[^1]

However, the slow `urbit play` functionality occurs for me on both my `macos-aarch64` and `linux-x86_64` machines, so I don't think this *particular* wart in our build configuration is the culprit. 

I will keep exploring tomorrow and report back here as I uncover anything of interest.

[^1]: 
```
Undefined symbols for architecture arm64:
  "_u3o_Config", referenced from:
      _main in main.o
      __cw_utils in main.o
      __cw_serf_writ in main.o
      __king_dawn in libvere.a(king.o)
      _u3_king_commence in libvere.a(king.o)
      __serf_grab in libvere.a(serf.o)
      _u3_serf_work in libvere.a(serf.o)
      ...
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by joemfb** [comment]
@matthew-levan which branch/commit are you building/running from, and with which pills? Are you booting and replaying with the same binary?

**Comment by matthew-levan** [comment]
@joemfb yesterday's builds were from `jb/moar-replay` using pills `1.16` through `1.19`, booted and replayed with the same binary (the one built from your branch).

**Comment by sigilante** [comment]
I have built two branches now, all essentially identical to `jb/moar-replay`.  They exhibit the same behavior.

**Comment by belisarius222** [comment]
@sigilante @joemfb What do you think is to be done about this PR, given that it is slow because of failed jet firings under certain conditions?  I'm guessing we should fix whatever that problem is before merging this, despite the fact that it doesn't seem like this PR is itself causing that issue.

**Comment by sigilante** [comment]
I'm of the opinion that we should not merge this until the root cause is fixed.  The different behavior on pills of various runtime versions is cursed.
